src/expr.jl

```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
     end
     return kwargs
 end
+
+"""
+    splitdef(ex::Expr; throw::Bool=true) -> Union{Dict{Symbol,Any}, Nothing}
+
+Split a function definition expression into its various parts including:
+
+- `:type`: Type of the function (long-form `:function`, short-form `:(=)`, anonymous `:(->)`)
+- `:name`: Name of the function (not present for anonymous functions)
+- `:params`: Parametric types defined on constructors
+- `:args`: Positional arguments of the function
+- `:kwargs`: Keyword arguments of the function
+- `:rtype`: Return type of the function
+- `:whereparams`: Where parameters
+- `:body`: Function body (not present for empty functions)
+
+All parts are optional with the exception of `:type`. When `throw=true` and the provided
+expression is not a function definition than an exception will be raised. When `throw=false`
+then `nothing` will be returned for invalid definitions.
+"""
+function splitdef(ex::Expr; throw::Bool=true)
+    def = Dict{Symbol,Any}()
```

Comment:
Why not a type or a NamedTuple?

Reply:
I would be good with returning a type. I ended up using a `Dict` to mostly keep compatibly with MacroTools' version